libgnucash/engine/Account.cpp

```diff
@@ -5617,24 +5591,21 @@ xaccAccountStagedTransactionTraversal (const Account *acc,
                                        void *cb_data)
 {
     AccountPrivate *priv;
-    GList *split_p;
-    GList *next;
     Transaction *trans;
-    Split *s;
     int retval;
 
     if (!acc) return 0;
 
     priv = GET_PRIVATE(acc);
-    for (split_p = priv->splits; split_p; split_p = next)
+    for (auto it = priv->splits.begin(); it != priv->splits.end();)
     {
         /* Get the next element in the split list now, just in case some
          * naughty thunk destroys the one we're using. This reduces, but
          * does not eliminate, the possibility of undefined results if
          * a thunk removes splits from this account. */
-        next = g_list_next(split_p);
+        auto next = std::next (it);
```

Comment:
That doesn't work for vectors. If a thunk destroys a split or moves it to a different account (either of which will remove it from `splits` behind our back) then `next` and all further iterators including `splits->end()` are invalid and continuing the loop would be undefined behavior. I think in practice what will happen is that we'd wind up skipping over the next split (because it's now pointed at by the current iterator and next points at the next one; vector iterators are just indexes) and then run off the end of the vector because end() points one split too far.

One safety measure would be to copy the vector and iterate on the copy, the other would be to set a flag on the Account and assert in `gnc_account_remove_split` that it's not set. The first is more in the spirit of the current design, though it conceivably could make for weird results. The second will crash. That might be an over-reaction depending on what's going on, but the backtrace will make it easy to figure out what happened.